pkg/daemon/metrics.go

```diff
@@ -0,0 +1,189 @@
+
+func RegisterMetrics(nodeName string) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(OffsetFromMaster)
+		prometheus.MustRegister(StateOfPI)
+		prometheus.MustRegister(FrequencyAdjustment)
+		prometheus.MustRegister(DelayFromMaster)
+
+		// Including these stats kills performance when Prometheus polls with multiple targets
+		prometheus.Unregister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+		prometheus.Unregister(prometheus.NewGoCollector())
+
+		NodeName = nodeName
+	})
+}
+
+// updatePTPMetrics ...
+func updatePTPMetrics(process string, offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster float64) {
+	OffsetFromMaster.With(prometheus.Labels{
+	"process": process,"node": NodeName}).Set(offsetFromMaster)
+
+	StateOfPI.With(prometheus.Labels{
+	"process": process,"node": NodeName}).Set(stateOfPI)
+
+	FrequencyAdjustment.With(prometheus.Labels{
+		"process": process,"node": NodeName}).Set(frequencyAdjustment)
+
+	DelayFromMaster.With(prometheus.Labels{
+		"process": process,"node": NodeName}).Set(delayFromMaster)
+}
+
+// extractMetrics ...
+func extractMetrics(processName, output string){
+	if processName == ptp4lProcessName && strings.Contains(output,"offset") {
+		offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster := extractPtp4lMetrics(output)
+		updatePTPMetrics(processName,offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster)
+	}
+
+	if processName == phc2sysProcessName && strings.Contains(output,"offset") {
+		offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster := extractPhc2sysMetrics(output)
+		updatePTPMetrics(processName,offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster)
+	}
+}
+
+func extractPtp4lMetrics(output string) (offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster float64) {
+	fields := strings.Fields(output)
+
+	if len(fields) != 10 {
+		glog.Error("ptp4l failed to extract metrics unknown output format")
+		return
+	}
+
+	offsetFromMaster, err := strconv.ParseFloat(fields[3], 64)
+	if err != nil {
+		glog.Errorf("ptp4l failed to parse offset from master output %s error %v",fields[3], err)
+	}
+
+	if len(fields[4]) != 2 {
+		glog.Errorf("ptp4l failed to parse state of pi output %s",fields[4])
+	}
+
+	stateOfPI, err = strconv.ParseFloat(string(fields[4][1]), 64)
+	if err != nil {
+		glog.Errorf("ptp4l failed to parse parse state of pi output %s error %v",string(fields[4][1]), err)
+	}
+
+	frequencyAdjustment, err = strconv.ParseFloat(fields[6], 64)
+	if err != nil {
+		glog.Errorf("ptp4l failed to parse frequency adjustment output %s error %v",fields[6], err)
+	}
+
+	delayFromMaster, err = strconv.ParseFloat(fields[9], 64)
+	if err != nil {
+		glog.Errorf("ptp4l failed to parse delay from master output %s error %v",fields[9], err)
+	}
+
+	return
+}
+
+func extractPhc2sysMetrics(output string) (offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster float64) {
+	fields := strings.Fields(output)
```

Comment:
Same as ptp4l,  `phc2sys` output has different format when `-u <summary-updates>` directive is appended.

With `summary-updates`:
```
phc2sys[20548.806]: CLOCK_REALTIME rms   40 max   70 freq -79866 +/-  14 delay   577 +/-   5
phc2sys[20553.807]: CLOCK_REALTIME rms   23 max   42 freq -79877 +/-  19 delay   565 +/-  31
phc2sys[20558.808]: CLOCK_REALTIME rms    4 max    7 freq -79872 +/-   3 delay   573 +/-   3
phc2sys[20563.809]: CLOCK_REALTIME rms   17 max   28 freq -79875 +/-  15 delay   563 +/-  29
```

Without `summary-updates`:
```
phc2sys[19785.802]: CLOCK_REALTIME phc offset         0 s2 freq  -79920 delay    572
phc2sys[19786.803]: CLOCK_REALTIME phc offset        15 s2 freq  -79905 delay    504
phc2sys[19787.803]: CLOCK_REALTIME phc offset         1 s2 freq  -79914 delay    579
phc2sys[19788.803]: CLOCK_REALTIME phc offset        -7 s2 freq  -79922 delay    505
phc2sys[19789.804]: CLOCK_REALTIME phc offset         0 s2 freq  -79917 delay    576
```